src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -517,6 +517,71 @@ public Activity AddEvent(ActivityEvent e)
             return this;
         }
 
+        /// <summary>
+        /// Add a <see cref="ActivityEvent" /> object containing the exception information to the <see cref="Events" /> list.
+        /// </summary>
+        /// <param name="exception">The exception to add to the attached events list.</param>
+        /// <param name="tags">The tags to add to the exception event.</param>
+        /// <param name="timestamp">The timestamp to add to the exception event.</param>
+        /// <returns><see langword="this" /> for convenient chaining.</returns>
+        /// <remarks>
+        /// - The name of the event will be "exception", and it will include the tags "exception.message", "exception.stacktrace", and "exception.type", in addition to the tags provided in the <paramref name="tags"/> parameter.
+        /// - Any registered <see cref="ActivityListener"/> with the <see cref="ActivityListener.ExceptionRecorder"/> callback will be notified about this exception addition before the <see cref="ActivityEvent" /> object is added to the <see cref="Events" /> list.
+        /// </remarks>
+        public Activity AddException(Exception exception, TagList tags = default, DateTimeOffset timestamp = default)
+        {
+            if (exception == null)
+            {
+                throw new ArgumentNullException(nameof(exception));
+            }
+
+            TagList exceptionTags = tags;
+
+            Source.NotifyActivityAddException(this, exception, ref exceptionTags);
+
+            const string exceptionEventName = "exception";
+            const string exceptionMessageTag = "exception.message";
+            const string exceptionStackTraceTag = "exception.stacktrace";
+            const string exceptionTypeTag = "exception.type";
+
+            bool hasMessage = false;
+            bool hasStackTrace = false;
+            bool hasType = false;
+
+            for (int i = 0; i < exceptionTags.Count; i++)
+            {
+                if (exceptionTags[i].Key == exceptionMessageTag)
+                {
+                    hasMessage = true;
+                }
+                else if (exceptionTags[i].Key == exceptionStackTraceTag)
+                {
+                    hasStackTrace = true;
+                }
+                else if (exceptionTags[i].Key == exceptionTypeTag)
+                {
+                    hasType = true;
+                }
+            }
+
+            if (!hasMessage)
+            {
+                exceptionTags.Add(new KeyValuePair<string, object?>(exceptionMessageTag, exception.Message));
+            }
+
+            if (!hasStackTrace)
+            {
+                exceptionTags.Add(new KeyValuePair<string, object?>(exceptionStackTraceTag, exception.ToString()));
```

Comment:
Should this be:
```suggestion
                exceptionTags.Add(new KeyValuePair<string, object?>(exceptionStackTraceTag, exception.StackTrace));
```
?

Reply:
We have decided to use `ToString()` for two reasons. First, it simplifies the stack trace by removing extra async frames https://github.com/microsoft/ApplicationInsights-dotnet/issues/2234#issuecomment-1249966175. Second, it includes nested stack traces. The end of this page: https://opentelemetry.io/docs/specs/semconv/exceptions/exceptions-spans/ specifying that.

Reply:
> First, it simplifies the stack trace by removing extra async frames

Exception.ToString() just uses Exception.StackTrace:
https://github.com/dotnet/runtime/blob/1ded19e700f267048a190fd45f949ed492469c1c/src/libraries/System.Private.CoreLib/src/System/Exception.cs#L130

What is the difference regarding extra async frames in the two cases? They should be identical in that regard.

> Second, it includes nested stack traces

Do you mean the stack traces of nested exceptions?  That's true, but that's also achievable by iterating through InnerException and including each nested exception's StackTrace.

> We have decided to use ToString() for two reasons

That's fine, it's just strange to me that the stack trace tag will contain duplicative information then for the message tag and the type tag, both of which are included in ToString.

Reply:
> The end of this page: https://opentelemetry.io/docs/specs/semconv/exceptions/exceptions-spans/ specifying that.

It would be great if the link is included in a comment so we will know why "exception.stacktrace" is `Exception.ToString()`.

Reply:
> > The end of this page: https://opentelemetry.io/docs/specs/semconv/exceptions/exceptions-spans/ specifying that.
> 
> It would be great if the link is included in a comment so we will know why "exception.stacktrace" is `Exception.ToString()`.

To be clear: that link is *not* the reason why `ToString()` is being used. It is totally up to implementations to decide what value to chose. The link lists one _possible approach_.

Reply:
Related: https://github.com/dotnet/coreclr/pull/14655

Reply:
> Related: [dotnet/coreclr#14655](https://github.com/dotnet/coreclr/pull/14655)

Yes. My point was those changes are in the implementation of Exception.StackTrace, which Exception.ToString then uses. The two stacks output here are identical:
```C#
try
{
    await MethodA();
}
catch (Exception ex)
{
    Console.WriteLine(ex.ToString());
    Console.WriteLine("==================================================================");
    Console.WriteLine(ex.StackTrace);
}

static async Task MethodA()
{
    await Task.Yield();
    await MethodB();
}

static async Task MethodB()
{
    await Task.Yield();
    await MethodC();
}

static async Task MethodC()
{
    await Task.Yield();
    await MethodD();
}

static async Task MethodD()
{
    await Task.Yield();
    throw new Exception("uh oh");
}
```

Reply:
Thanks, @stephentoub, for the clarification. I prefer to continue using `ToString()` as it already includes the nested exception stack traces, unlike manually enumerating and formatting them. Including the duplicated information when using `ToString()` is not a significant issue.